File: soc-server/internal/services/alert_service.go
// ==============================================================================
// Package services - Alert Service
// File: internal/services/alert_service.go
// Mô tả: Business logic cho Alert management.
//         Cung cấp các thao tác CRUD, filter, và lifecycle transitions.
// ==============================================================================

package services

import (
	"context"
	"fmt"
	"log"
	"strings"
	"time"

	"soc-server/internal/models"
	"soc-server/pkg/database"

	"gorm.io/gorm"
)

// AlertService - Service xử lý business logic cho Alerts
type AlertService struct {
	db    *gorm.DB
	redis *database.RedisClient
}

func stringPointer(value string) *string {
	return &value
}

// DefaultAgents returns a small set of demo agents so the dashboard looks populated immediately.
func DefaultAgents() []models.Agent {
	return []models.Agent{
		{
			ID:                  "agent-demo-win-01",
			Hostname:            "WIN-CLIENT-01",
			IPAddress:           "10.0.0.21",
			OSType:              "windows",
			Status:              "online",
			LastHeartbeatAt:     &[]time.Time{time.Now().Add(-2 * time.Minute)}[0],
			MTLSCertFingerprint: stringPointer("demo-win-01-cert"),
			AgentVersion:        "1.4.2",
		},
		{
			ID:                  "agent-demo-linux-01",
			Hostname:            "LINUX-SRV-07",
			IPAddress:           "10.0.0.42",
			OSType:              "linux",
			Status:              "online",
			LastHeartbeatAt:     &[]time.Time{time.Now().Add(-3 * time.Minute)}[0],
			MTLSCertFingerprint: stringPointer("demo-linux-01-cert"),
			AgentVersion:        "1.4.2",
		},
	}
}

// DefaultAlerts returns demo SOC alerts for first-run dashboard rendering.
func DefaultAlerts() []models.Alert {
	return []models.Alert{
		{
			ID:               "alert-demo-001",
			AgentID:          "agent-demo-win-01",
			RuleID:           "RULE-POWERSHELL-ENCODED",
			EventType:        models.EventSuspiciousProcess,
			Severity:         models.SeverityHigh,
			Status:           models.AlertStatusNew,
			Title:            "PowerShell encoded command execution",
			Description:      "PowerShell executed an encoded payload with suspicious flags commonly used by malware loaders.",
			MITRETactic:      "Execution",
			MITRETechniqueID: "T1059.001",
			RawPayload:       `{"process_name":"powershell.exe","command_line":"powershell -EncodedCommand JABjAGwAaQBlAG4...","host":"WIN-CLIENT-01"}`,
		},
		{
			ID:               "alert-demo-002",
			AgentID:          "agent-demo-linux-01",
			RuleID:           "RULE-RANSOMWARE-VSS",
			EventType:        models.EventNetworkAnomaly,
			Severity:         models.SeverityCritical,
			Status:           models.AlertStatusInProgress,
			Title:            "Shadow copy deletion attempt",
			Description:      "Endpoint attempted to delete local shadow copies in a pattern consistent with ransomware behavior.",
			MITRETactic:      "Impact",
			MITRETechniqueID: "T1490",
			RawPayload:       `{"process_name":"vssadmin.exe","command_line":"vssadmin.exe shadow delete /all /quiet","host":"LINUX-SRV-07"}`,
		},
		{
			ID:               "alert-demo-003",
			AgentID:          "agent-demo-win-01",
			RuleID:           "RULE-NETWORK-C2",
			EventType:        models.EventNetworkAnomaly,
			Severity:         models.SeverityMedium,
			Status:           models.AlertStatusResolved,
			Title:            "Outbound suspicious connection",
			Description:      "New outbound TCP connection to an external IP outside the known internal subnets.",
			MITRETactic:      "Command and Control",
			MITRETechniqueID: "T1071",
			RawPayload:       `{"dst_ip":"185.220.101.182","protocol":"tcp","host":"WIN-CLIENT-01"}`,
		},
	}
}

// NewAlertService - Khởi tạo AlertService
func NewAlertService(db *gorm.DB, redis *database.RedisClient) *AlertService {
	return &AlertService{db: db, redis: redis}
}

// GetAllAlerts - Lấy danh sách tất cả alerts với phân trang
// Hỗ trợ filter theo severity, status, event_type, agent_id
func (s *AlertService) GetAllAlerts(page, pageSize int, filters map[string]string) ([]models.Alert, int64, error) {
	var alerts []models.Alert
	var total int64

	query := s.db.Model(&models.Alert{})
	// system_metric là telemetry định kỳ, không phải security alert trên dashboard.
	query = query.Where("event_type <> ?", "system_metric")

	// Áp dụng filters
	if severity, ok := filters["severity"]; ok && severity != "" {
		query = query.Where("severity = ?", severity)
	}
	if status, ok := filters["status"]; ok && status != "" {
		query = query.Where("status = ?", status)
	}
	if eventType, ok := filters["event_type"]; ok && eventType != "" {
		query = query.Where("event_type = ?", eventType)
	}
	if agentID, ok := filters["agent_id"]; ok && agentID != "" {
		query = query.Where("agent_id = ?", agentID)
	}

	// Đếm tổng số records
	query.Count(&total)

	// Phân trang và sắp xếp theo thời gian mới nhất
	offset := (page - 1) * pageSize
	err := query.Order("created_at DESC").
		Offset(offset).Limit(pageSize).
		Preload("Agent"). // Preload thông tin Agent
		Find(&alerts).Error

	return alerts, total, err
}

// GetAlertByID - Lấy alert theo ID
func (s *AlertService) GetAlertByID(id string) (*models.Alert, error) {
	var alert models.Alert
	err := s.db.Preload("Agent").First(&alert, "id = ?", id).Error
	if err != nil {
		return nil, fmt.Errorf("không tìm thấy Alert '%s': %w", id, err)
	}
	return &alert, nil
}

// CreateAlert - Tạo alert mới
func (s *AlertService) CreateAlert(alert *models.Alert) error {
	return s.db.Create(alert).Error
}

// UpdateAlertStatus - Cập nhật trạng thái Alert (dùng nội bộ)
func (s *AlertService) UpdateAlertStatus(id string, status models.AlertStatus) error {
	result := s.db.Model(&models.Alert{}).Where("id = ?", id).Update("status", status)
	if result.RowsAffected == 0 {
		return fmt.Errorf("không tìm thấy Alert '%s'", id)
	}
	return result.Error
}

// UpdateAlertStatusWithContext - Cập nhật trạng thái + context từ n8n SOAR
// Description: kết quả xử lý SOAR (markdown)
// Tags: nhãn phân loại (ai-decision, auto-blocked, soar...)
func (s *AlertService) UpdateAlertStatusWithContext(id string, status models.AlertStatus, description string, tags []string) error {
	updates := map[string]interface{}{
		"status": status,
	}
	if description != "" {
		updates["description"] = description
	}
	// Tags lưu dưới dạng JSON string trong description nếu model chưa có cột tags riêng
	// Nếu bạn muốn lưu tags riêng, thêm cột tags VARCHAR vào model Alert

	result := s.db.Model(&models.Alert{}).Where("id = ?", id).Updates(updates)
	if result.RowsAffected == 0 {
		return fmt.Errorf("không tìm thấy Alert '%s'", id)
	}
	return result.Error
}

// GetAlertStats - Thống kê alerts cho Dashboard
func (s *AlertService) GetAlertStats() (map[string]interface{}, error) {
	stats := make(map[string]interface{})
	ctx := context.Background()

	// Tổng số alert trong 24h
	start24h := time.Now().Add(-24 * time.Hour)
	var totalAlertsToday int64
	s.db.Model(&models.Alert{}).Where("created_at >= ?", start24h).Count(&totalAlertsToday)
	stats["total_alerts_today"] = totalAlertsToday

	// Alert critical trong 24h
	var criticalAlerts int64
	s.db.Model(&models.Alert{}).Where("severity = ? AND created_at >= ?", models.SeverityCritical, start24h).Count(&criticalAlerts)
	stats["critical_alerts"] = criticalAlerts

	// Tổng agent / online agent
	var agentsTotal int64
	s.db.Model(&models.Agent{}).Count(&agentsTotal)
	stats["agents_total"] = agentsTotal

	agentsOnline := int64(0)
	if s.redis != nil {
		var agentIDs []string
		if err := s.db.Model(&models.Agent{}).Pluck("id", &agentIDs).Error; err == nil {
			for _, agentID := range agentIDs {
				isOnline, err := s.redis.IsAgentOnline(ctx, agentID)
				if err == nil && isOnline {
					agentsOnline++
				}
			}
		}
	} else {
		s.db.Model(&models.Agent{}).Where("status = ?", "online").Count(&agentsOnline)
	}
	stats["agents_online"] = agentsOnline

	// Active cases
	var activeCases int64
	s.db.Model(&models.Case{}).Where("status != ? AND status != ?", models.CaseStatusClosed, models.CaseStatusRejected).Count(&activeCases)
	stats["active_cases"] = activeCases

	// Alert trend last 24h — 1 query duy nhất thay vì 72 queries
	type HourlyRow struct {
		Hour     string
		Severity string
		Count    int64
	}
	var rows []HourlyRow
	start24h2 := time.Now().Add(-24 * time.Hour)

	// MySQL: DATE_FORMAT + GROUP BY để đếm theo giờ xá severity
	s.db.Raw(`
		SELECT DATE_FORMAT(created_at, '%H:00') AS hour,
		       severity,
		       COUNT(*) AS count
		FROM alerts
		WHERE created_at >= ?
		  AND severity IN ('critical','high','medium')
		GROUP BY hour, severity
		ORDER BY hour ASC
	`, start24h2).Scan(&rows)

	// Chuyển kết quả thành map[hour][severity]=count
	hourMap := make(map[string]map[string]int64)
	for _, r := range rows {
		if _, ok := hourMap[r.Hour]; !ok {
			hourMap[r.Hour] = map[string]int64{"critical": 0, "high": 0, "medium": 0}
		}
		hourMap[r.Hour][r.Severity] = r.Count
	}

	// Tạo chuỗi 24 slots theo thứ tự thời gian
	alertTrend := make([]map[string]interface{}, 0, 24)
	for i := 23; i >= 0; i-- {
		hourLabel := time.Now().Add(-time.Duration(i+1) * time.Hour).Truncate(time.Hour).Format("15:00")
		counts := hourMap[hourLabel]
		alertTrend = append(alertTrend, map[string]interface{}{
			"hour":     hourLabel,
			"critical": counts["critical"],
			"high":     counts["high"],
			"medium":   counts["medium"],
		})
	}
	stats["alert_trend"] = alertTrend

	// Severity distribution
	severityMap := map[string]int64{}
	for _, severity := range []models.AlertSeverity{models.SeverityCritical, models.SeverityHigh, models.SeverityMedium, models.SeverityLow} {
		var count int64
		s.db.Model(&models.Alert{}).Where("severity = ?", severity).Count(&count)
		severityMap[string(severity)] = count
	}
	severityDistribution := []map[string]interface{}{
		{"name": "Critical", "value": severityMap[string(models.SeverityCritical)], "color": "#ff3366"},
		{"name": "High", "value": severityMap[string(models.SeverityHigh)], "color": "#ff9900"},
		{"name": "Medium", "value": severityMap[string(models.SeverityMedium)], "color": "#eab308"},
		{"name": "Low", "value": severityMap[string(models.SeverityLow)], "color": "#60a5fa"},
	}
	stats["severity_distribution"] = severityDistribution

	// Top affected agents
	type TopAgent struct {
		Hostname   string
		AlertCount int64
	}
	var topAgents []TopAgent
	s.db.Model(&models.Alert{}).Select("agents.hostname as hostname, count(*) as alert_count").
		Joins("LEFT JOIN agents ON agents.id = alerts.agent_id").
		Group("alerts.agent_id").
		Order("alert_count DESC").
		Limit(5).Scan(&topAgents)
	formattedTopAgents := make([]map[string]interface{}, 0, len(topAgents))
	for _, agent := range topAgents {
		formattedTopAgents = append(formattedTopAgents, map[string]interface{}{
			"hostname":    strings.TrimSpace(agent.Hostname),
			"alert_count": agent.AlertCount,
		})
	}
	stats["top_agents"] = formattedTopAgents

	// Preserve legacy keys used elsewhere in other code
	stats["total"] = totalAlertsToday
	stats["by_severity"] = severityMap
	stats["by_status"] = map[string]int64{"new": 0, "in_progress": 0, "resolved": 0}

	log.Printf("[ALERT SERVICE] Dashboard stats: totalAlertsToday=%d, agentsOnline=%d/%d, activeCases=%d", totalAlertsToday, agentsOnline, agentsTotal, activeCases)
	return stats, nil
}

// ==============================================================================
// StartAlertCleanupJob - Job tự động dọn dẹp Alert cũ (chạy background goroutine)
//
// Logic:
//   - Chạy mỗi 24 giờ một lần
//   - Xóa Alert có severity = 'low' hoặc rule_id = 'EVENT-OBSERVED'
//     nếu created_at cũ hơn retentionDays ngày
//   - Alert HIGH/CRITICAL/MEDIUM được giữ lại vĩnh viễn
//
// Gọi hàm này một lần trong main() với `go alertService.StartAlertCleanupJob(ctx, retentionDays)`
// ==============================================================================
func (s *AlertService) StartAlertCleanupJob(ctx context.Context, retentionDays int) {
	if retentionDays <= 0 {
		retentionDays = 7
	}
	ticker := time.NewTicker(24 * time.Hour)
	defer ticker.Stop()

	log.Printf("[CLEANUP JOB] 🧹 Khởi động Alert Cleanup Job: xóa LOW alerts cũ hơn %d ngày", retentionDays)

	// Chạy ngay lần đầu khi khởi động
	s.runCleanup(retentionDays)

	for {
		select {
		case <-ticker.C:
			s.runCleanup(retentionDays)
		case <-ctx.Done():
			log.Println("[CLEANUP JOB] Dừng Alert Cleanup Job.")
			return
		}
	}
}

// runCleanup - Thực hiện xóa một lần
func (s *AlertService) runCleanup(retentionDays int) {
	cutoff := time.Now().AddDate(0, 0, -retentionDays)
	result := s.db.Where(
		"(severity = ? OR rule_id = ?) AND created_at < ?",
		string(models.SeverityLow), "EVENT-OBSERVED", cutoff,
	).Delete(&models.Alert{})

	if result.Error != nil {
		log.Printf("[CLEANUP JOB] ❌ Lỗi khi dọn dẹp alert cũ: %v", result.Error)
		return
	}
	if result.RowsAffected > 0 {
		log.Printf("[CLEANUP JOB] ✅ Đã xóa %d LOW/EVENT-OBSERVED alerts cũ hơn %d ngày",
			result.RowsAffected, retentionDays)
	}
}

File: soc-server-python/app/services/alert_service.py
# ==============================================================================
# app/services/alert_service.py - Alert Service
# Tương đương: internal/services/alert_service.go
# ==============================================================================

import asyncio
import logging
import uuid
from datetime import datetime, timedelta
from typing import Optional, Tuple, List, Dict, Any

from sqlalchemy import select, func, desc, delete, text
from sqlalchemy.orm import joinedload
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.alert import Alert, AlertSeverity, AlertStatus
from app.models.agent import Agent
from app.models.case import Case

logger = logging.getLogger(__name__)


def default_agents() -> List[dict]:
    return [
        {
            "id": "agent-demo-win-01",
            "hostname": "WIN-CLIENT-01",
            "ip_address": "10.0.0.21",
            "os_type": "windows",
            "status": "online",
            "last_heartbeat_at": datetime.utcnow() - timedelta(minutes=2),
            "mtls_cert_fingerprint": "demo-win-01-cert",
            "agent_version": "1.4.2",
        },
        {
            "id": "agent-demo-linux-01",
            "hostname": "LINUX-SRV-07",
            "ip_address": "10.0.0.42",
            "os_type": "linux",
            "status": "online",
            "last_heartbeat_at": datetime.utcnow() - timedelta(minutes=3),
            "mtls_cert_fingerprint": "demo-linux-01-cert",
            "agent_version": "1.4.2",
        },
    ]


def default_alerts() -> List[dict]:
    return [
        {
            "id": "alert-demo-001",
            "agent_id": "agent-demo-win-01",
            "rule_id": "RULE-POWERSHELL-ENCODED",
            "event_type": "suspicious_process",
            "severity": "high",
            "status": "new",
            "title": "PowerShell encoded command execution",
            "description": "PowerShell executed an encoded payload with suspicious flags commonly used by malware loaders.",
            "mitre_tactic": "Execution",
            "mitre_technique_id": "T1059.001",
            "raw_payload": '{"process_name":"powershell.exe","command_line":"powershell -EncodedCommand JABjAGwAaQBlAG4...","host":"WIN-CLIENT-01"}',
        },
        {
            "id": "alert-demo-002",
            "agent_id": "agent-demo-linux-01",
            "rule_id": "RULE-RANSOMWARE-VSS",
            "event_type": "network_anomaly",
            "severity": "critical",
            "status": "in_progress",
            "title": "Shadow copy deletion attempt",
            "description": "Endpoint attempted to delete local shadow copies in a pattern consistent with ransomware behavior.",
            "mitre_tactic": "Impact",
            "mitre_technique_id": "T1490",
            "raw_payload": '{"process_name":"vssadmin.exe","command_line":"vssadmin.exe shadow delete /all /quiet","host":"LINUX-SRV-07"}',
        },
        {
            "id": "alert-demo-003",
            "agent_id": "agent-demo-win-01",
            "rule_id": "RULE-NETWORK-C2",
            "event_type": "network_anomaly",
            "severity": "medium",
            "status": "resolved",
            "title": "Outbound suspicious connection",
            "description": "New outbound TCP connection to an external IP outside the known internal subnets.",
            "mitre_tactic": "Command and Control",
            "mitre_technique_id": "T1071",
            "raw_payload": '{"dst_ip":"185.220.101.182","protocol":"tcp","host":"WIN-CLIENT-01"}',
        },
    ]


class AlertService:
    def __init__(self, db: AsyncSession, redis_client=None):
        self.db = db
        self.redis = redis_client

    async def get_all_alerts(
        self,
        page: int = 1,
        page_size: int = 100,
        filters: Optional[Dict[str, str]] = None,
    ) -> Tuple[List[Alert], int]:
        """Lấy danh sách tất cả alerts với phân trang và eager load Agent"""
        filters = filters or {}
        stmt = select(Alert).options(joinedload(Alert.agent)).where(Alert.event_type != "system_metric")
        count_stmt = select(func.count(Alert.id)).where(Alert.event_type != "system_metric")

        if filters.get("severity"):
            stmt = stmt.where(Alert.severity == filters["severity"])
            count_stmt = count_stmt.where(Alert.severity == filters["severity"])
        if filters.get("status"):
            stmt = stmt.where(Alert.status == filters["status"])
            count_stmt = count_stmt.where(Alert.status == filters["status"])
        if filters.get("event_type"):
            stmt = stmt.where(Alert.event_type == filters["event_type"])
            count_stmt = count_stmt.where(Alert.event_type == filters["event_type"])
        if filters.get("agent_id"):
            stmt = stmt.where(Alert.agent_id == filters["agent_id"])
            count_stmt = count_stmt.where(Alert.agent_id == filters["agent_id"])

        total_res = await self.db.execute(count_stmt)
        total = total_res.scalar() or 0

        offset = (page - 1) * page_size
        stmt = stmt.order_by(desc(Alert.created_at)).offset(offset).limit(page_size)

        result = await self.db.execute(stmt)
        alerts = list(result.scalars().all())
        return alerts, total

    async def get_alert_by_id(self, alert_id: str) -> Optional[Alert]:
        """Lấy alert theo ID với Agent preload"""
        stmt = select(Alert).options(joinedload(Alert.agent)).where(Alert.id == alert_id)
        res = await self.db.execute(stmt)
        return res.scalar_one_or_none()

    async def create_alert(self, data: Dict[str, Any]) -> Alert:
        """Tạo alert mới"""
        alert = Alert(
            id=data.get("id") or str(uuid.uuid4()),
            agent_id=data["agent_id"],
            rule_id=data.get("rule_id"),
            event_type=data["event_type"],
            severity=data["severity"],
            raw_payload=data.get("raw_payload", "{}"),
            status=data.get("status", "new"),
            title=data.get("title", ""),
            description=data.get("description", ""),
            mitre_tactic=data.get("mitre_tactic"),
            mitre_technique_id=data.get("mitre_technique_id"),
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        self.db.add(alert)
        await self.db.commit()
        await self.db.refresh(alert)
        return alert

    async def update_alert_status(self, alert_id: str, status: str) -> bool:
        """Cập nhật trạng thái Alert"""
        stmt = select(Alert).where(Alert.id == alert_id)
        res = await self.db.execute(stmt)
        alert = res.scalar_one_or_none()
        if not alert:
            return False

        alert.status = status
        alert.updated_at = datetime.utcnow()
        await self.db.commit()
        return True

    async def update_alert_status_with_context(
        self,
        alert_id: str,
        status: str,
        description: str = "",
        tags: Optional[List[str]] = None,
    ) -> bool:
        """Cập nhật trạng thái + context từ n8n SOAR"""
        stmt = select(Alert).where(Alert.id == alert_id)
        res = await self.db.execute(stmt)
        alert = res.scalar_one_or_none()
        if not alert:
            return False

        alert.status = status
        if description:
            alert.description = description
        alert.updated_at = datetime.utcnow()
        await self.db.commit()
        return True

    async def get_alert_stats(self) -> Dict[str, Any]:
        """Thống kê alerts cho Dashboard"""
        stats: Dict[str, Any] = {}
        now = datetime.utcnow()
        start24h = now - timedelta(hours=24)

        # 1. Tổng alerts trong 24h
        res_today = await self.db.execute(
            select(func.count(Alert.id)).where(Alert.created_at >= start24h)
        )
        total_alerts_today = res_today.scalar() or 0
        stats["total_alerts_today"] = total_alerts_today

        # 2. Phân bố severity
        severity_map = {"critical": 0, "high": 0, "medium": 0, "low": 0}
        for sev in ["critical", "high", "medium", "low"]:
            cnt_res = await self.db.execute(
                select(func.count(Alert.id)).where(Alert.severity == sev)
            )
            severity_map[sev] = cnt_res.scalar() or 0

        stats["critical_alerts"] = severity_map["critical"]
        stats["severity_distribution"] = [
            {"name": "Critical", "value": severity_map["critical"], "color": "#ff3366"},
            {"name": "High", "value": severity_map["high"], "color": "#ff9900"},
            {"name": "Medium", "value": severity_map["medium"], "color": "#eab308"},
            {"name": "Low", "value": severity_map["low"], "color": "#60a5fa"},
        ]

        # 3. Agents total & online
        cnt_agents = await self.db.execute(select(func.count(Agent.id)))
        stats["agents_total"] = cnt_agents.scalar() or 0

        cnt_online = await self.db.execute(select(func.count(Agent.id)).where(Agent.status == "online"))
        stats["agents_online"] = cnt_online.scalar() or 0

        # 4. Active cases
        cnt_cases = await self.db.execute(
            select(func.count(Case.id)).where(Case.status.not_in(["Closed", "Rejected"]))
        )
        stats["active_cases"] = cnt_cases.scalar() or 0

        # 5. Alert trend last 24h
        trend_query = text("""
            SELECT DATE_FORMAT(created_at, '%H:00') AS hour,
                   severity,
                   COUNT(*) AS count
            FROM alerts
            WHERE created_at >= :start24h
              AND severity IN ('critical', 'high', 'medium')
            GROUP BY hour, severity
            ORDER BY hour ASC
        """)
        trend_rows = await self.db.execute(trend_query, {"start24h": start24h})
        hour_map = {}
        for row in trend_rows.fetchall():
            hour_str, sev, count = row[0], row[1], row[2]
            if hour_str not in hour_map:
                hour_map[hour_str] = {"critical": 0, "high": 0, "medium": 0}
            hour_map[hour_str][sev] = count

        alert_trend = []
        for i in range(23, -1, -1):
            target_hour = (now - timedelta(hours=i)).strftime("%H:00")
            counts = hour_map.get(target_hour, {"critical": 0, "high": 0, "medium": 0})
            alert_trend.append({
                "hour": target_hour,
                "critical": counts.get("critical", 0),
                "high": counts.get("high", 0),
                "medium": counts.get("medium", 0),
            })
        stats["alert_trend"] = alert_trend

        # 6. Top affected agents
        top_agents_query = text("""
            SELECT COALESCE(agents.hostname, 'Unknown') AS hostname, COUNT(*) AS alert_count
            FROM alerts
            LEFT JOIN agents ON agents.id = alerts.agent_id
            GROUP BY alerts.agent_id, agents.hostname
            ORDER BY alert_count DESC
            LIMIT 5
        """)
        top_agents_rows = await self.db.execute(top_agents_query)
        top_agents = [
            {"hostname": row[0].strip() if row[0] else "Unknown", "alert_count": row[1]}
            for row in top_agents_rows.fetchall()
        ]
        stats["top_agents"] = top_agents

        # Legacy keys
        stats["total"] = total_alerts_today
        stats["by_severity"] = severity_map
        stats["by_status"] = {"new": 0, "in_progress": 0, "resolved": 0}

        return stats

    async def run_cleanup(self, retention_days: int = 7):
        """Xóa LOW/EVENT-OBSERVED alerts cũ hơn N ngày"""
        cutoff = datetime.utcnow() - timedelta(days=retention_days)
        stmt = delete(Alert).where(
            ((Alert.severity == "low") | (Alert.rule_id == "EVENT-OBSERVED"))
            & (Alert.created_at < cutoff)
        )
        res = await self.db.execute(stmt)
        await self.db.commit()
        if res.rowcount:
            logger.info(f"[CLEANUP JOB] ✅ Đã xóa {res.rowcount} LOW alerts cũ hơn {retention_days} ngày")
